mcstatus/motd.py

```diff
@@ -0,0 +1,577 @@
+
+    They're just removing unused formatting, unaffected colors etc.
+    """
+
+    def __init__(self, raw: list[ParsedMotdItem]) -> None:
+        self.parsed: list[ParsedMotdItem] = raw
+
+    def get_simplifies(self) -> Iterable[Callable[[], set[int]]]:
+        """Generator with all simplifies.
+
+        They must follow this signature:
+        ``(self) -> set[int]``
+        Where ``int`` is an index of the element, which need to be removed.
+        """
+        for item in dir(self):
+            if callable(getattr(self, item)) and not item.startswith("_") and item != "get_simplifies":
+                yield getattr(self, item)
+
+    def remove_double_items(self) -> set[int]:
+        """Get indices of all doubled items that can be removed."""
+        to_remove: set[int] = set()
+
+        for index, item in enumerate(self.parsed):
+            try:
+                next_item = self.parsed[index + 1]
+            except IndexError:  # Last item (without any next item)
+                break
+
+            if item == next_item:
+                to_remove.add(index)
+
+        return to_remove
+        
+    def remove_double_colors(self) -> set[int]:
+        """Get indices of all doubled color items.
+
+        As colors (obviously) override each other, we only ever care about the last one, ignore
+        the previous ones. (for example: specifying red color, then orange, then yellow, then some text
+        will just result in yellow text)
+        """
+        to_remove: set[int] = set()
+
+        prev_color: int | None = None
+        for index, item in enumerate(self.parsed):
+            if isinstance(item, (MinecraftColor, WebColor)):
+                # If we found a color after another, remove the previous color
+                if prev_color is not None:
+                    to_remove.add(prev_color)
+                prev_color = index
+
+            # If we find a string, that's what our color we found previously applies to,
+            # set prev_color to None, marking this color as used
+            if isinstance(item, str):
+                prev_color = None
+
+        return to_remove
+
+    def remove_unused_formatting_before_color(self) -> set[int]:
+        """Remove unused formatting before color.
+
+        As colors (obviously) override formatting, we only ever care about the color, ignore
+        the formatting. (for example: specifying bold formatting, then italic, then yellow, then some text
+        will just result in yellow text)
+        """
+        to_remove: set[int] = set()
+        
+        for index, item in enumerate(self.parsed):
+            try:
+                next_item = self.parsed[index + 1]
+            except IndexError:  # Last item (without any next item)
+                break
+
+            if isinstance(item, Formatting) and isinstance(next_item, (MinecraftColor, WebColor)):
+                to_remove.add(index)
+    
+        return to_remove
```

Comment:
With the new logic, this code will only run once. It isn't enough to just check whether the next item is a color, and remove the formatting directly before it. What if there were 2 formatting specifications (like: bold, italic) and a color afterwards. There's no while loop to keep rerunning these, meaning you need to handle this all in this function now.

To keep this at O(n), you can use this solution:
```suggestion
        to_remove: set[int] = set()
        
        collected_formattings = []
        for index, item in enumerate(self.parsed):
            # Collect the indices of all formatting items
            if isinstance(item, Formatting):
                collected_formattings.apped(index)
                continue
                
            # Only proceed if we already have some formatting items collected
            if len(collected_formattings) == 0:
                continue
                
            # If we found a string (before a color), then that's what the formatting(s) apply to,
            # remove them from collected items as they're used up here
            if isinstance(item, str):
                collected_formattings = []
                continue
            
            # If we found a color (before text), then our collected formatting items will be overridden
            # and are useless. Mark them for removal.
            if isinstance(item, (MinecraftColor, WebColor)):
                for fmt_index in collected_formattings:
                    to_remove.add(fmt_index)
                collected_formattings = []
                
        return to_remove
```

Reply:
Fixed [here](https://github.com/py-mine/mcstatus/compare/5a99136ba6ccfe68f2ad9ac32efeb5742a8ef097..33969fa91ce672da8406b30686151ffb9113ab62).